File: StkExtensionPlugins/Graphics/VB.Net/ProjectionRasterStreamPlugin/ProjectionRasterStreamPlugin.vb
Imports System.Collections.Generic
Imports System.Runtime.InteropServices
Imports System.Text
Imports AGI.STKGraphics.Plugins

<Guid("6DE9FCA4-95F0-41d2-BE84-46CFF18767EB")> _
<ClassInterface(ClassInterfaceType.None)> _
<ProgId("ProjectionRasterStreamPlugin.VBNET")> _
Public Class ProjectionRasterStreamPlugin
    Implements IAgStkGraphicsPluginProjectionStream
    Implements IAgStkGraphicsPluginRasterStream
    Public Sub New()
        ' Set default values
        NearPlane = 1.0
        FarPlane = 10000
        FieldOfViewHorizontal = 0.5
        FieldOfViewVertical = 0.5
        Dates = New List(Of Double)()
        Positions = New List(Of Array)()
        Orientations = New List(Of Array)()
    End Sub

#Region "IAgStkGraphicsPluginProjectionStream Members"

    Private Function IAgStkGraphicsPluginProjectionStream_OnGetFirstProjection(ByVal Time As AGI.STKUtil.IAgDate, ByVal pContext As IAgStkGraphicsPluginProjectionStreamContext) As Boolean Implements IAgStkGraphicsPluginProjectionStream.OnGetFirstProjection
        pContext.NearPlane = NearPlane
        pContext.FarPlane = FarPlane
        pContext.FieldOfViewHorizontal = FieldOfViewHorizontal
        pContext.FieldOfViewVertical = FieldOfViewVertical

        Dim result As ProjectionPositionOrientation = EvaluateProjectionPositionOrientation(Double.Parse(Time.Format("epSec")), 0, Dates.Count)
        Dim xyz As Array = result.Position
        Dim quat As Array = result.Orientation

        pContext.SetPosition(xyz)
        pContext.SetOrientation(quat)

        Return True
    End Function

    Private Function IAgStkGraphicsPluginProjectionStream_OnGetNextProjection(ByVal Time As AGI.STKUtil.IAgDate, ByVal NextTime As AGI.STKUtil.IAgDate, ByVal pContext As IAgStkGraphicsPluginProjectionStreamContext) As Boolean Implements IAgStkGraphicsPluginProjectionStream.OnGetNextProjection
        pContext.NearPlane = NearPlane
        pContext.FarPlane = FarPlane
        pContext.FieldOfViewHorizontal = FieldOfViewHorizontal
        pContext.FieldOfViewVertical = FieldOfViewVertical

        Dim result As ProjectionPositionOrientation = EvaluateProjectionPositionOrientation(Double.Parse(Time.Format("epSec")), 0, Dates.Count)
        Dim xyz As Array = result.Position
        Dim quat As Array = result.Orientation

        pContext.SetPosition(xyz)
        pContext.SetOrientation(quat)

        Return True
    End Function

#End Region

#Region "IAgStkGraphicsPluginRasterStream Members"

    Private Function IAgStkGraphicsPluginRasterStream_OnGetRasterAttributes(ByVal pAttributes As IAgStkGraphicsPluginRasterStreamAttributes) As Boolean Implements IAgStkGraphicsPluginRasterStream.OnGetRasterAttributes
        ' Set gifProvider for the raster
        gifProvider = New GifProvider(RasterPath)

        ' Assign the raster's attributes
        pAttributes.Width = gifProvider.Size.Width
        pAttributes.Height = gifProvider.Size.Height
        pAttributes.RasterFormat = AgEStkGraphicsPluginRasterFormat.eStkGraphicsPluginRasterFormatBgra
        pAttributes.RasterType = AgEStkGraphicsPluginRasterType.eStkGraphicsPluginRasterTypeUnsignedByte

        Return True
    End Function

    Private Function IAgStkGraphicsPluginRasterStream_OnGetNextRaster(ByVal Time As AGI.STKUtil.IAgDate, ByVal NextTime As AGI.STKUtil.IAgDate, ByVal pContext As IAgStkGraphicsPluginRasterStreamContext) As Boolean Implements IAgStkGraphicsPluginRasterStream.OnGetNextRaster
        If m_LastTime Is Nothing Then
            m_LastTime = Time.Subtract("sec", 1.0)
        End If

        ' Only update if animating
        If Time.OLEDate > m_LastTime.OLEDate Then
            pContext.RasterAsBitmap.SetBitmap(gifProvider.NextFrame().GetHbitmap())
            m_LastTime = Time
        ElseIf Time.OLEDate < m_LastTime.OLEDate Then
            pContext.RasterAsBitmap.SetBitmap(gifProvider.PreviousFrame().GetHbitmap())
            m_LastTime = Time
        End If

        Return True
    End Function

#End Region

    Private Function EvaluateProjectionPositionOrientation(ByVal searchDate As Double, ByVal startIndex As Integer, ByVal searchLength As Integer) As ProjectionPositionOrientation
        ' Find the midpoint of the length
        Dim midpoint As Integer = startIndex + (searchLength \ 2)

        ' Base cases
        If Dates(startIndex) = searchDate Then
            Return New ProjectionPositionOrientation(Positions(startIndex), Orientations(startIndex))
        End If
        If Dates(midpoint) = searchDate Then
            Return New ProjectionPositionOrientation(Positions(midpoint), Orientations(midpoint))
        End If
        If searchLength = 3 Then
            If searchDate < Dates(midpoint) Then
                Return InterpolatePositionAndOrientation(midpoint - 1, midpoint, searchDate)
            Else
                Return InterpolatePositionAndOrientation(midpoint, midpoint + 1, searchDate)
            End If
        End If
        If searchLength = 2 Then
            InterpolatePositionAndOrientation(startIndex, startIndex + 1, searchDate)
        End If
        If searchLength < 2 Then
            ' This will only occur if animating backwards, so try to interpolate with the lower index first.
            If startIndex - 1 > 0 Then
                Return InterpolatePositionAndOrientation(startIndex, startIndex + 1, searchDate)
            ElseIf startIndex + 1 < Dates.Count Then
                Return InterpolatePositionAndOrientation(startIndex - 1, startIndex, searchDate)
            Else
                Return New ProjectionPositionOrientation(Positions(startIndex), Orientations(startIndex))
            End If
        End If

        ' Normal case: binary search
        If searchDate < Dates(midpoint) Then
            Return EvaluateProjectionPositionOrientation(searchDate, startIndex, midpoint - startIndex)
        Else
            Return EvaluateProjectionPositionOrientation(searchDate, midpoint + 1, startIndex + searchLength - (midpoint + 1))
        End If
    End Function

    ''' <summary>
    ''' Interpolates a ProjectionPositionOrientation that would suffice at a <paramref name="desiredDate"/> between
    ''' Dates[<paramref name="lowerIndex"/>] and Dates[<paramref name="higherIndex"/>].
    ''' </summary>
    ''' <param name="lowerIndex">The index reprenting the position in Dates that serves as the lower bound of interpolation.</param>
    ''' <param name="higherIndex">The index reprenting the position in Dates that serves as the upper bound of interpolation.</param>
    ''' <param name="desiredDate">The date to interpolate for.</param>
    ''' <returns>A ProjectionPositionOrientation containing the interpolated position and orientation.</returns>
    Private Function InterpolatePositionAndOrientation(ByVal lowerIndex As Integer, ByVal higherIndex As Integer, ByVal desiredDate As Double) As ProjectionPositionOrientation
        Dim range As Double = Dates(higherIndex) - Dates(lowerIndex)
        Dim locationInRange As Double = desiredDate - Dates(lowerIndex)
        Dim ratio As Double = locationInRange / range

        Return New ProjectionPositionOrientation(InterpolatePosition(Positions(lowerIndex), Positions(higherIndex), ratio), InterpolateOrientation(Orientations(lowerIndex), Orientations(higherIndex), ratio))
    End Function

    ''' <summary>
    ''' This function will calculate an interpolated cartesian position according to the linear interpolation (LERP) formula
    ''' such that its location is an amount <paramref name="t"/> between <paramref name="pos1"/> and <paramref name="pos2"/>.
    ''' </summary>
    ''' <param name="pos1">The first IAgCartesian3Vector. Serves as the lower position of the interpolation.</param>
    ''' <param name="pos2">The second IAgCartesian3Vector. Serves as the uppoer position of the interpolation.</param>
    ''' <param name="t">The amount of interpolation (i.e. the desired amount between <paramref name="pos1"/> and <paramref name="pos2"/> to calculate).</param>
    ''' <returns>An Array of cartesian position components that is an amount <paramref name="t"/> between <paramref name="pos1"/> and <paramref name="pos2"/>.</returns>
    Private Shared Function InterpolatePosition(ByVal pos1 As Array, ByVal pos2 As Array, ByVal t As Double) As Array
        Return New Object() {CDbl(pos1.GetValue(0)) + (CDbl(pos2.GetValue(0)) - CDbl(pos1.GetValue(0))) * t, CDbl(pos1.GetValue(1)) + (CDbl(pos2.GetValue(1)) - CDbl(pos1.GetValue(1))) * t, CDbl(pos1.GetValue(2)) + (CDbl(pos2.GetValue(2)) - CDbl(pos1.GetValue(2))) * t}
    End Function

    ''' <summary>
    ''' This function will calculate an interpolated quaternion according to the spherical linear interpolation (SLERP) formula
    ''' such that its values lay an amount <paramref name="t"/> between <paramref name="q1"/> and <paramref name="q2"/>.
    ''' </summary>
    ''' <param name="q1">The first IAgOrientation. Serves as the lower quaterion of the interpolation.</param>
    ''' <param name="q2">the second IAgOrientation. Serves as the upper quaterion of the interpolation.</param>
    ''' <param name="t">The amount of interpolation (i.e. the desired amount between <paramref name="q1"/> and <paramref name="q1"/> to calculate).</param>
    ''' <returns>An Array of quaterion components that lay an amount <paramref name="t"/> between <paramref name="q1"/> and <paramref name="q2"/>.</returns>
    Private Shared Function InterpolateOrientation(ByVal q1 As Array, ByVal q2 As Array, ByVal t As Double) As Array
        Dim EPS As Double = 0.001

        Dim theta As Double, cosineTheta As Double, sineTheta As Double, scale0 As Double, scale1 As Double
        Dim qi As Double() = New Double(3) {}

        Dim q1x As Double = CDbl(q1.GetValue(0)), q1y As Double = CDbl(q1.GetValue(1)), q1z As Double = CDbl(q1.GetValue(2)), q1w As Double = CDbl(q1.GetValue(3))

        Dim q2x As Double = CDbl(q2.GetValue(0)), q2y As Double = CDbl(q2.GetValue(1)), q2z As Double = CDbl(q2.GetValue(2)), q2w As Double = CDbl(q2.GetValue(3))

        ' Do a linear interpolation between two quaternions (0 <= t <= 1).
        cosineTheta = q1x * q2x + q1y * q2y + q1z * q2z + q1w * q2w
        ' dot product
        If cosineTheta < 0 Then
            cosineTheta = -cosineTheta
            qi(0) = -q2x
            qi(1) = -q2y
            qi(2) = -q2z
            qi(3) = -q2w
        Else
            qi(0) = q2x
            qi(1) = q2y
            qi(2) = q2z
            qi(3) = q2w
        End If

        If (1 - cosineTheta) <= EPS Then
            ' If the quaternions are really close, do a simple linear interpolation.
            scale0 = 1 - t
            scale1 = t
        Else
            ' Otherwise SLERP.
            theta = CSng(Math.Acos(cosineTheta))
            sineTheta = CSng(Math.Sin(theta))
            scale0 = CSng(Math.Sin((1 - t) * theta)) / sineTheta
            scale1 = CSng(Math.Sin(t * theta)) / sineTheta
        End If

        ' Calculate interpolated quaternion:
        Dim resultX As Double, resultY As Double, resultZ As Double, resultW As Double
        resultX = scale0 * q1x + scale1 * qi(0)
        resultY = scale0 * q1y + scale1 * qi(1)
        resultZ = scale0 * q1z + scale1 * qi(2)
        resultW = scale0 * q1w + scale1 * qi(3)

        Return New Object() {resultX, resultY, resultZ, resultW}
    End Function

    Public Property NearPlane() As Double
        Get
            Return m_NearPlane
        End Get
        Set(ByVal value As Double)
            m_NearPlane = value
        End Set
    End Property
    Private m_NearPlane As Double
    Public Property FarPlane() As Double
        Get
            Return m_FarPlane
        End Get
        Set(ByVal value As Double)
            m_FarPlane = value
        End Set
    End Property
    Private m_FarPlane As Double
    Public Property FieldOfViewHorizontal() As Double
        Get
            Return m_FieldOfViewHorizontal
        End Get
        Set(ByVal value As Double)
            m_FieldOfViewHorizontal = value
        End Set
    End Property
    Private m_FieldOfViewHorizontal As Double
    Public Property FieldOfViewVertical() As Double
        Get
            Return m_FieldOfViewVertical
        End Get
        Set(ByVal value As Double)
            m_FieldOfViewVertical = value
        End Set
    End Property
    Private m_FieldOfViewVertical As Double
    Public Property Dates() As IList(Of Double)
        Get
            Return m_Dates
        End Get
        Set(ByVal value As IList(Of Double))
            m_Dates = value
        End Set
    End Property
    Private m_Dates As IList(Of Double)
    Public Property Positions() As IList(Of Array)
        Get
            Return m_Positions
        End Get
        Set(ByVal value As IList(Of Array))
            m_Positions = value
        End Set
    End Property
    Private m_Positions As IList(Of Array)
    Public Property Orientations() As IList(Of Array)
        Get
            Return m_Orientations
        End Get
        Set(ByVal value As IList(Of Array))
            m_Orientations = value
        End Set
    End Property
    Private m_Orientations As IList(Of Array)

    Public Property RasterPath() As String
        Get
            Return m_RasterPath
        End Get
        Set(ByVal value As String)
            m_RasterPath = value
        End Set
    End Property
    Private m_RasterPath As String
    Private gifProvider As GifProvider
    Private m_LastTime As AGI.STKUtil.IAgDate
End Class

File: StkExtensionPlugins/Graphics/CSharp/ProjectionRasterStreamPlugin/ProjectionRasterStreamPlugin.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Text;
using AGI.STKGraphics.Plugins;
using Microsoft.Win32;

namespace ProjectionRasterStreamPlugin
{
	[Guid("5B1AC88E-3ADA-437c-9F40-1B8184ECB485")]
    [ClassInterface(ClassInterfaceType.None)]
    [ProgId("ProjectionRasterStreamPlugin.CSharp")]
    public class ProjectionRasterStreamPlugin : IAgStkGraphicsPluginProjectionStream,
                                                IAgStkGraphicsPluginRasterStream
    {
        public ProjectionRasterStreamPlugin()
        {
            // Set default values
            NearPlane = 1.0;
            FarPlane = 10000;
            FieldOfViewHorizontal = 0.5;
            FieldOfViewVertical = 0.5;
            Dates = new List<double>();
            Positions = new List<Array>();
            Orientations = new List<Array>();
        }

        #region IAgStkGraphicsPluginProjectionStream Members

        bool IAgStkGraphicsPluginProjectionStream.OnGetFirstProjection(AGI.STKUtil.IAgDate Time, IAgStkGraphicsPluginProjectionStreamContext pContext)
        {
            pContext.NearPlane = NearPlane;
            pContext.FarPlane = FarPlane;
            pContext.FieldOfViewHorizontal = FieldOfViewHorizontal;
            pContext.FieldOfViewVertical = FieldOfViewVertical;

            ProjectionPositionOrientation result = EvaluateProjectionPositionOrientation(double.Parse(Time.Format("epSec")), 0, Dates.Count);
            Array xyz = result.Position;
            Array quat = result.Orientation;

            pContext.SetPosition(ref xyz);
            pContext.SetOrientation(ref quat);

            return true;
        }

        bool IAgStkGraphicsPluginProjectionStream.OnGetNextProjection(AGI.STKUtil.IAgDate Time, AGI.STKUtil.IAgDate NextTime, IAgStkGraphicsPluginProjectionStreamContext pContext)
        {
            pContext.NearPlane = NearPlane;
            pContext.FarPlane = FarPlane;
            pContext.FieldOfViewHorizontal = FieldOfViewHorizontal;
            pContext.FieldOfViewVertical = FieldOfViewVertical;

            ProjectionPositionOrientation result = EvaluateProjectionPositionOrientation(double.Parse(Time.Format("epSec")), 0, Dates.Count);
            Array xyz = result.Position;
            Array quat = result.Orientation;

            pContext.SetPosition(ref xyz);
            pContext.SetOrientation(ref quat);

            return true;
        }

        #endregion

        #region IAgStkGraphicsPluginRasterStream Members

        bool IAgStkGraphicsPluginRasterStream.OnGetRasterAttributes(IAgStkGraphicsPluginRasterStreamAttributes pAttributes)
        {
            // Set gifProvider for the raster
            gifProvider = new GifProvider(RasterPath);

            // Assign the raster's attributes
            pAttributes.Width = gifProvider.Size.Width;
            pAttributes.Height = gifProvider.Size.Height;
            pAttributes.RasterFormat = AgEStkGraphicsPluginRasterFormat.eStkGraphicsPluginRasterFormatBgra;
            pAttributes.RasterType = AgEStkGraphicsPluginRasterType.eStkGraphicsPluginRasterTypeUnsignedByte;

            return true;
        }

        bool IAgStkGraphicsPluginRasterStream.OnGetNextRaster(AGI.STKUtil.IAgDate Time, AGI.STKUtil.IAgDate NextTime, IAgStkGraphicsPluginRasterStreamContext pContext)
        {
            if (m_LastTime == null)
                m_LastTime = Time.Subtract("sec", 1.0);

            // Only update if animating
            if (Time.OLEDate > m_LastTime.OLEDate)
            {
                pContext.RasterAsBitmap.SetBitmap(gifProvider.NextFrame().GetHbitmap());
                m_LastTime = Time;
            }
            else if (Time.OLEDate < m_LastTime.OLEDate)
            {
                pContext.RasterAsBitmap.SetBitmap(gifProvider.PreviousFrame().GetHbitmap());
                m_LastTime = Time;
            }

            return true;
        }

        #endregion

        private ProjectionPositionOrientation EvaluateProjectionPositionOrientation(double searchDate, int startIndex, int searchLength)
        {
            // Find the midpoint of the length
            int midpoint = startIndex + (searchLength / 2);

            // Base cases
            if (Dates[startIndex] == searchDate)
            {
                return new ProjectionPositionOrientation(Positions[startIndex], Orientations[startIndex]);
            }
            if (Dates[midpoint] == searchDate)
            {
                return new ProjectionPositionOrientation(Positions[midpoint], Orientations[midpoint]);
            }
            if (searchLength == 3)
            {
                if (searchDate < Dates[midpoint])
                    return InterpolatePositionAndOrientation(midpoint - 1, midpoint, searchDate);
                else
                    return InterpolatePositionAndOrientation(midpoint, midpoint + 1, searchDate);
            }
            if (searchLength == 2)
            {
                InterpolatePositionAndOrientation(startIndex, startIndex + 1, searchDate);
            }
            if (searchLength < 2)
            {
                // This will only occur if animating backwards, so try to interpolate with the lower index first.
                if (startIndex - 1 > 0)
                    return InterpolatePositionAndOrientation(startIndex, startIndex + 1, searchDate);
                else if (startIndex + 1 < Dates.Count)
                    return InterpolatePositionAndOrientation(startIndex - 1, startIndex, searchDate);
                else
                    return new ProjectionPositionOrientation(Positions[startIndex], Orientations[startIndex]);
            }

            // Normal case: binary search
            if (searchDate < Dates[midpoint])
            {
                return EvaluateProjectionPositionOrientation(searchDate, startIndex, midpoint - startIndex);
            }
            else
            {
                return EvaluateProjectionPositionOrientation(searchDate, midpoint + 1, startIndex + searchLength - (midpoint + 1));
            }
        }

        /// <summary>
        /// Interpolates a <typeparamref name="ProjectionPositionOrientation"/> that would suffice at a <paramref name="desiredDate"/> between
        /// Dates[<paramref name="lowerIndex"/>] and Dates[<paramref name="higherIndex"/>].
        /// </summary>
        /// <param name="lowerIndex">The index reprenting the position in Dates that serves as the lower bound of interpolation.</param>
        /// <param name="higherIndex">The index reprenting the position in Dates that serves as the upper bound of interpolation.</param>
        /// <param name="desiredDate">The date to interpolate for.</param>
        /// <returns>A <typeparamref name="ProjectionPositionOrientation"/> containing the interpolated position and orientation.</returns>
        private ProjectionPositionOrientation InterpolatePositionAndOrientation(int lowerIndex, int higherIndex, double desiredDate)
        {
            double range = Dates[higherIndex] - Dates[lowerIndex];
            double locationInRange = desiredDate - Dates[lowerIndex];
            double ratio = locationInRange / range;

            return new ProjectionPositionOrientation(
                InterpolatePosition(Positions[lowerIndex], Positions[higherIndex], ratio),
                InterpolateOrientation(Orientations[lowerIndex], Orientations[higherIndex], ratio));
        }

        /// <summary>
        /// This function will calculate an interpolated cartesian position according to the linear interpolation (LERP) formula
        /// such that its location is an amount <paramref name="t"/> between <paramref name="pos1"/> and <paramref name="pos2"/>.
        /// </summary>
        /// <param name="pos1">The first IAgCartesian3Vector. Serves as the lower position of the interpolation.</param>
        /// <param name="pos2">The second IAgCartesian3Vector. Serves as the uppoer position of the interpolation.</param>
        /// <param name="t">The amount of interpolation (i.e. the desired amount between <paramref name="pos1"/> and <paramref name="pos2"/> to calculate).</param>
        /// <returns>An <typeparamref name="Array"/> of cartesian position components that is an amount <paramref name="t"/> between <paramref name="pos1"/> and <paramref name="pos2"/>.</returns>
        private static Array InterpolatePosition(Array pos1, Array pos2, double t)
        {
            return new object[]
            {
                (double)pos1.GetValue(0) + ((double)pos2.GetValue(0) - (double)pos1.GetValue(0)) * t,
                (double)pos1.GetValue(1) + ((double)pos2.GetValue(1) - (double)pos1.GetValue(1)) * t,
                (double)pos1.GetValue(2) + ((double)pos2.GetValue(2) - (double)pos1.GetValue(2)) * t,
            };
        }

        /// <summary>
        /// This function will calculate an interpolated quaternion according to the spherical linear interpolation (SLERP) formula
        /// such that its values lay an amount <paramref name="t"/> between <paramref name="q1"/> and <paramref name="q2"/>.
        /// </summary>
        /// <param name="q1">The first IAgOrientation. Serves as the lower quaterion of the interpolation.</param>
        /// <param name="q2">the second IAgOrientation. Serves as the upper quaterion of the interpolation.</param>
        /// <param name="t">The amount of interpolation (i.e. the desired amount between <paramref name="q1"/> and <paramref name="q1"/> to calculate).</param>
        /// <returns>An <typeparamref name="Array"/> of quaterion components that lay an amount <paramref name="t"/> between <paramref name="q1"/> and <paramref name="q2"/>.</returns>
        private static Array InterpolateOrientation(Array q1, Array q2, double t)
        {
            double EPS = 0.001;

            double theta, cosineTheta, sineTheta, scale0, scale1;
            double[] qi = new double[4];

            double q1x = (double)q1.GetValue(0),
                   q1y = (double)q1.GetValue(1),
                   q1z = (double)q1.GetValue(2),
                   q1w = (double)q1.GetValue(3);

            double q2x = (double)q2.GetValue(0),
                   q2y = (double)q2.GetValue(1),
                   q2z = (double)q2.GetValue(2),
                   q2w = (double)q2.GetValue(3);

            // Do a linear interpolation between two quaternions (0 <= t <= 1).
            cosineTheta = q1x * q2x + q1y * q2y + q1z * q2z + q1w * q2w;  // dot product

            if (cosineTheta < 0)
            {
                cosineTheta = -cosineTheta;
                qi[0] = -q2x;
                qi[1] = -q2y;
                qi[2] = -q2z;
                qi[3] = -q2w;
            }
            else
            {
                qi[0] = q2x;
                qi[1] = q2y;
                qi[2] = q2z;
                qi[3] = q2w;
            }

            if ((1 - cosineTheta) <= EPS)// If the quaternions are really close, do a simple linear interpolation.
            {
                scale0 = 1 - t;
                scale1 = t;
            }
            else
            {
                // Otherwise SLERP.
                theta = (float)Math.Acos(cosineTheta);
                sineTheta = (float)Math.Sin(theta);
                scale0 = (float)Math.Sin((1 - t) * theta) / sineTheta;
                scale1 = (float)Math.Sin(t * theta) / sineTheta;
            }

            // Calculate interpolated quaternion:
            double resultX, resultY, resultZ, resultW;
            resultX = scale0 * q1x + scale1 * qi[0];
            resultY = scale0 * q1y + scale1 * qi[1];
            resultZ = scale0 * q1z + scale1 * qi[2];
            resultW = scale0 * q1w + scale1 * qi[3];

            return new object[] { resultX, resultY, resultZ, resultW };
        }

        #region Registration functions
        /// <summary>
        /// Called when the assembly is registered for use from COM.
        /// </summary>
        /// <param name="t">The type being exposed to COM.</param>
        [ComRegisterFunction]
        [ComVisible(false)]
        public static void RegisterFunction(Type t)
        {
            RemoveOtherVersions(t);
        }

        /// <summary>
        /// Called when the assembly is unregistered for use from COM.
        /// </summary>
        /// <param name="t">The type exposed to COM.</param>
        [ComUnregisterFunctionAttribute]
        [ComVisible(false)]
        public static void UnregisterFunction(Type t)
        {
            // Do nothing.
        }

        /// <summary>
        /// Called when the assembly is registered for use from COM.
        /// Eliminates the other versions present in the registry for
        /// this type.
        /// </summary>
        /// <param name="t">The type being exposed to COM.</param>
        public static void RemoveOtherVersions(Type t)
        {
            try
            {
                using (RegistryKey clsidKey = Registry.ClassesRoot.OpenSubKey("CLSID"))
                {
                    StringBuilder guidString = new StringBuilder("{");
                    guidString.Append(t.GUID.ToString());
                    guidString.Append("}");
                    using (RegistryKey guidKey = clsidKey.OpenSubKey(guidString.ToString()))
                    {
                        if (guidKey != null)
                        {
                            using (RegistryKey inproc32Key = guidKey.OpenSubKey("InprocServer32", true))
                            {
                                if (inproc32Key != null)
                                {
                                    string currentVersion = t.Assembly.GetName().Version.ToString();
                                    string[] subKeyNames = inproc32Key.GetSubKeyNames();
                                    if (subKeyNames.Length > 1)
                                    {
                                        foreach (string subKeyName in subKeyNames)
                                        {
                                            if (subKeyName != currentVersion)
                                            {
                                                inproc32Key.DeleteSubKey(subKeyName);
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
            catch
            {
                // Ignore all exceptions...
            }
        }
        #endregion

        public double NearPlane { get; set; }
        public double FarPlane { get; set; }
        public double FieldOfViewHorizontal { get; set; }
        public double FieldOfViewVertical { get; set; }
        public IList<double> Dates { get; set; }
        public IList<Array> Positions { get; set; }
        public IList<Array> Orientations { get; set; }

        public string RasterPath { get; set; }
        private GifProvider gifProvider;
        private AGI.STKUtil.IAgDate m_LastTime;
    }
}
